### ISSUE TITLE
[PROPOSAL] Publish protobuf files.

### ISSUE THREAD

**Issue by dblock** [opensearch-project/opensearch-api-specification#532]
### What are you proposing?

Coming from https://github.com/opensearch-project/OpenSearch/pull/15411 and https://github.com/opensearch-project/OpenSearch/issues/15190 I propose we generate and publish protobuf files as part of the [build/release workflow](https://github.com/opensearch-project/opensearch-api-specification/blob/main/.github/workflows/build.yml).

### What users have asked for this feature?

OpenSearch is trying to use the OpenAPI spec to generate .proto files then use those on both client and server sides. 

### What problems are you trying to solve?

OpenSearch server wants .proto files.
OpenSearch clients that will implement gRPC/protobuf want .proto files.

### What is the developer experience going to be?

A release of opensearch-api-specification will contain a .zip with .proto files.

### Why should it be built? Any reason not to?

The right place for .proto files is this repo because multiple consumers want .proto files.

**Comment by dblock** [comment]
cc: @andrross @amberzsy

**Comment by philiplhchan** [comment]
I think currently one challenge for an automated release pipeline is how we're going to ensure that the generated proto definitions are backwards compatible, e.g. if we add a new field/type in the API spec, the new generated proto should not change any of the field tags, type names, etc. of any of the existing definitions because protobuf doesn't allow that. 

1. Ideally the transpiler should be stable (e.g. it wouldn't work if the tag IDs are generated randomly on each execution)
2. we'd need some automated check to ensure no previously defined/published types/fields are modified, otherwise it would break the API for any clusters/clients running different versions.

**Comment by dblock** [comment]
@philiplhchan I think you're saying we need to generate proto files for a given API version. This can be done by first generating a version specific API (`npm run merge -- --opensearch-version=2.16`, see [doc](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md#spec-merger)). What this does is that it removes everything per semver. Then you can covert that to .proto files. Right now the main pre-release is a moving target as we're fixing the spec quickly, but obviously a real release could follow this.

Regardless, consumers of the spec should not rely on a moving target. Commit the output and update it via PRs so you can catch anything breaking.

**Comment by amberzsy** [comment]
as we are moving forward to have certain automation and transpiler tooling for specs -> proto. need to have certain criteria and rules. 

1. Muti-data types: [[1](https://github.com/opensearch-project/opensearch-api-specification/blob/19421f502740967e4d6df102f1c5765c53eaa010/spec/schemas/_core.rank_eval.yaml#L159)],[[2](https://github.com/opensearch-project/opensearch-api-specification/blob/19421f502740967e4d6df102f1c5765c53eaa010/spec/schemas/_common.yaml#L149)]
example: 
```
 type: 
        - boolean
        - 'null'
        - number
        - object
        - string
OR 
type: ['null', number, string]
```
proposal: 
```
oneof <property_name> {
  string string_value = 1;
  bool boolean_value = 2;
  bytes object = 3;
 ....
}
```

2. `oneOf` for Object and Array of same Object
```
      oneOf:
        - $ref: '#/components/schemas/DataStreamName'
        - type: array
          items:
            $ref: '#/components/schemas/DataStreamName'

```
proposal: 
simplified with 
```
- type: array
          items:
            $ref: '#/components/schemas/DataStreamName'
```
so in proto it will be defined as `repeated <Defined_Property>;`

3. Missing definition for Object. esp for DSL query type. 
```
 TermsQuery:
      allOf:
        - $ref: '#/components/schemas/QueryBase'
        - type: object

   MatchAllQuery:
      allOf:
        - $ref: '#/components/schemas/QueryBase'
        - type: object
```
for Query, we might need to have certain definition. for non Query, we can use `google/protobuf/any. `

4. Knn in the request body but not in dsl query which slightly different layout in JSON

in json: 
```
{
  "size": 2,
  "query": {
    "knn": {
      "my_vector2": {
        "vector": [2, 3, 5, 6],
        "k": 2
      }
    }
  }
}
```
in Specs:  it's in request body but not dsl specs define [here](https://github.com/opensearch-project/opensearch-api-specification/blob/19421f502740967e4d6df102f1c5765c53eaa010/spec/namespaces/_core.yaml#L2437), should it yet another Query type? 
 
5. in proto3, it removed required/optional, it will make it hard to check presence of primitives. as discussed [here](https://stackoverflow.com/questions/51918871/check-if-a-field-has-been-set-in-protocol-buffer-3)
one solution is to have all primitives use [wrapper proto](https://github.com/protocolbuffers/protobuf/blob/main/src/google/protobuf/wrappers.proto).  

6. most of numeric number has def as "number", either we introduce a  General Number wrapper type or have more strict type defined in Specs. 

7. There are many Enum definition. which some of them are duplicated. e.g Dutch defined twice in [Language](https://github.com/protocolbuffers/protobuf/blob/main/src/google/protobuf/wrappers.proto) and [DutchAnalyzer](https://github.com/opensearch-project/opensearch-api-specification/blob/19421f502740967e4d6df102f1c5765c53eaa010/spec/schemas/_common.analysis.yaml#L345). it will cause duplicate enum error when generating proto. By default, the generator tool will have Enum defined as 

```
enum DutchAnalyzerEnum {
      TypeDutchAnalyzerEnum_Dutch = 1;
}

enum LanguageEnum {
    TypeLanguageEnum_Dutch = 1;
}
```
we can either follow ^^ or .. change specs 

7. some enum type defined as [[example](https://github.com/opensearch-project/opensearch-api-specification/blob/19421f502740967e4d6df102f1c5765c53eaa010/spec/schemas/indices._common.yaml#L197)]
```
    IndexCheckOnStartup:
      type: string
      enum:
        - checksum
        - 'false'
        - 'true'
```

should we convert to 
```
    IndexCheckOnStartup:
      type: string
      enum:
        - checksum
        - FALSE
        - TRUE
```

**Comment by amberzsy** [comment]
Question/Alignment regarding protobuf files format. 

Q1: how should we handle proto message hierarchy. currently the openapi-generator tool create new proto files for each object, which it end up with tons of files. Should we want to group into few proto files? 

proposal: 
1. all toplevel dsl need to be in same file, otherwise it will run into recursive dependency issue. 
2. all request body have one (same for response)
3. all request params have one file (or we can have 2 and 3 in same file)
4. some proto can move to common.proto. 

Q2: how we decode the inner fields. 
e.g in [aggregation specs](https://github.com/opensearch-project/opensearch-api-specification/blob/main/spec/schemas/_common.aggregations.yaml#L801)

by default, it will decode all the way the innerest fields def and flatten into current proto. 

```
message ChildrenAggregate {
        <TYPE> agg_base_field_x = 1;
        <TYPE> agg_base_field_y = 2;
        <TYPE> agg_base_field_z = 3;
         ..... 
}
```
however, it won't work when there's recursive definition as it will have duplicate field. 
should we flatten and remove duplicates or have 
```
message ChildrenAggregate {
        <AggrBase> agg_base = 1;
         ..... 
}
message AggrBase {
        <TYPE> field_x = 1;
        <TYPE> field_y = 2;
        <TYPE> field_z = 3;
         ..... 
}
```

**Comment by dblock** [comment]
> a) can we follow same format here? easier to parsing.

Yes. We would want a linter to fail when something inconsistent is used, also @nhtruong might have an opinion here.

**Comment by dblock** [comment]
> Should we want to group into few proto files?

To me what's important is to organize the source code that is written by humans well. But these are machine generated so I wouldn't, it's less work and easier to grok changes over smaller files than merged ones.

**Comment by amberzsy** [comment]
all proto published to https://github.com/opensearch-project/opensearch-protobufs. @dblock we can close this.
